Source/WebKit/UIProcess/Automation/win/WebAutomationSessionWin.cpp

```diff
@@ -0,0 +1,365 @@
+
+    ASSERT_NOT_REACHED();
+    return 0;
+}
+
+static int keyCodeForCharKey(CharKey charKey)
+{
+    if ('A' <= charKey && charKey <= 'Z')
+        return VK_A + (charKey - 'A');
+    if ('a' <= charKey && charKey <= 'z')
+        return VK_A + (charKey - 'a');
+    if ('0' <= charKey && charKey <= '9')
+        return VK_0 + (charKey - '0');
+
+    // TODO: other symbols (!"#$%&'()=~|`@[]+*;:<>,./\?_ etc...)
+    // ASSERT_NOT_REACHED();
+    return 0;
+}
+
+static unsigned modifiersForKeyCode(unsigned keyCode)
+{
+    switch (keyCode) {
+    case VK_SHIFT: // just in case
+    case VK_LSHIFT:
+    case VK_RSHIFT:
+        return kWKEventModifiersShiftKey;
+    case VK_CONTROL: // just in case
+    case VK_LCONTROL:
+    case VK_RCONTROL:
+        return kWKEventModifiersControlKey;
+    case VK_MENU: // just in case
+    case VK_LMENU:
+    case VK_RMENU:
+        return kWKEventModifiersAltKey;
+    }
+    return 0;
+}
+
+void WebAutomationSession::platformSimulateKeyboardInteraction(WebPageProxy& page, KeyboardInteraction interaction, std::variant<VirtualKey, CharKey>&& key)
+{
+    int keyCode;
+    WTF::switchOn(key,
+        [&](VirtualKey virtualKey) {
+            keyCode = keyCodeForVirtualKey(virtualKey);
+        },
+        [&](CharKey charKey) {
+            keyCode = keyCodeForCharKey(charKey);
+        }
+    );
+
+    unsigned modifiers = modifiersForKeyCode(keyCode);
+
+    UINT message;
+    WPARAM wparam = keyCode;
+    LPARAM lparam = 0; // FIXME
+
+    switch (interaction) {
+    case KeyboardInteraction::KeyPress:
+        m_currentModifiers |= modifiers;
+        message = WM_KEYDOWN;
+        break;
+    case KeyboardInteraction::KeyRelease:
+        m_currentModifiers &= ~modifiers;
+        message = WM_KEYUP;
+        break;
+    case KeyboardInteraction::InsertByKey:
+        message = WM_KEYDOWN;
+        break;
+    }
+
+    auto hwnd = reinterpret_cast<HWND>(page.viewWidget());
+    NativeWebKeyboardEvent event(hwnd, message, wparam, lparam, { });
+    page.handleKeyboardEvent(event);
+}
+
+OptionSet<WebEventModifier> WebAutomationSession::platformWebModifiersFromRaw(unsigned modifiers)
+{
+    OptionSet<WebEventModifier> webModifiers;
+
+    if (modifiers & kWKEventModifiersAltKey)
+        webModifiers.add(WebEventModifier::AltKey);
+    if (modifiers & kWKEventModifiersControlKey)
+        webModifiers.add(WebEventModifier::ControlKey);
+    if (modifiers & kWKEventModifiersShiftKey)
+        webModifiers.add(WebEventModifier::ShiftKey);
+
+    return webModifiers;
+}
+
+void WebAutomationSession::platformSimulateKeySequence(WebPageProxy& page, const String& keySequence)
+{
+    // https://www.w3.org/TR/uievents-code/#key-alphanumeric-writing-system
+    String keyIdentifier = "Key"_s + keySequence.convertToASCIIUppercase();
+
+    auto hwnd = reinterpret_cast<HWND>(page.viewWidget());
+    String key = (m_currentModifiers & kWKEventModifiersShiftKey)? keySequence.convertToASCIIUppercase(): keySequence;
```

Comment:
Not documented, but I think you have to add white spaces before `:` and `?` in WebKit..  

`x ? a : b`, not `x? a: b`.

Reply:
Documented here! https://webkit.org/code-style-guidelines/#spacing-binary-ternary-op

Reply:
I missed it because the `Tools/Scripts/check-webkit-style` did not make an error.
Thanks for pointing this out.